drivers/flash/flash_mspi_infineon_s25h.c

```diff
@@ -0,0 +1,837 @@
+		return ret;
+	}
+
+	ret = flash_mspi_infineon_s25h_verify_jedec_id(dev);
+	if (ret < 0) {
+		LOG_ERR("JEDEC ID mismatch after switching to full quad MSPI mode. Communication "
+			"is broken");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int flash_mspi_infineon_s25h_disable_hybrid_sector_mode(const struct device *dev)
+{
+	/* This driver needs the hybrid sector mode to be disabled. So if it's found to be turned on
+	 * it gets changed. This requires changing the non-volatile configuration and also a reset
+	 */
+	int ret = 0;
+	uint8_t conf3 = 0;
+
+	ret = flash_mspi_infineon_s25h_rw_any_register(dev, INF_MSPI_S25H_ADDRESS_VOLATILE_CFG_3,
+						       &conf3, 0, MSPI_RX);
+	if (ret < 0) {
+		LOG_ERR("Error reading volatile configuration register 3");
+		return ret;
+	}
+
+	if ((conf3 & INF_MSPI_S25H_CFG_3_UNHYSA_BIT) == 0) {
+		LOG_INF("Flash is in hybrid sector mode. Changing non-volatile config to correct "
+			"this");
+
+		conf3 |= INF_MSPI_S25H_CFG_3_UNHYSA_BIT;
+
+		ret = flash_mspi_infineon_s25h_rw_any_register(
+			dev, INF_MSPI_S25H_ADDRESS_NON_VOLATILE_CFG_3, &conf3, 0, MSPI_TX);
+		if (ret < 0) {
+			LOG_ERR("Error changing non-volatile configuration of flash");
+			return ret;
+		}
+
+		ret = flash_mspi_infineon_s25h_wait_for_idle(dev,
+							     INF_MSPI_S25H_TIMEOUT_IDLE_STARTUP);
+		if (ret < 0) {
+			LOG_ERR("Error waiting for flash to enter idle after disabling hybrid "
+				"sector mode");
+			return ret;
+		}
+
+		ret = flash_mspi_infineon_s25h_reset(dev);
+		if (ret < 0) {
+			LOG_ERR("Error resetting flash via reset command");
+			return ret;
+		}
+
+		conf3 = 0;
+		ret = flash_mspi_infineon_s25h_rw_any_register(
+			dev, INF_MSPI_S25H_ADDRESS_VOLATILE_CFG_3, &conf3, 0, MSPI_RX);
+		if (ret < 0) {
+			LOG_ERR("Error reading volatile config 3 register of flash");
+			return ret;
+		}
+
+		if ((conf3 & INF_MSPI_S25H_CFG_3_UNHYSA_BIT) == 0) {
+			LOG_ERR("Changing the flash configuration to Uniform mode didn't work");
+			return -EIO;
+		}
+
+		ret = flash_mspi_infineon_s25h_set_writing_forbidden(dev, true);
+		if (ret < 0) {
+			LOG_ERR("Error re-enabling the write protection");
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int flash_mspi_infineon_s25h_enter_4_byte_address_mode(const struct device *dev)
+{
+	int ret = 0;
+	const struct flash_mspi_infineon_s25h_cfg *config = dev->config;
+	struct flash_mspi_infineon_s25h_data *data = dev->data;
+
+	const struct mspi_xfer_packet enter_4_byte_cmd = {
+		.dir = MSPI_TX,
+		.cmd = INF_MSPI_S25H_OPCODE_ENABLE_4_BYTE_ADDR_MODE,
+		.num_bytes = 0,
+	};
+
+	struct mspi_xfer xfer = {
+		INF_MSPI_S25H_DEFAULT_XFER_DATA,
+		.rx_dummy = 0,
+		.addr_length = 0,
+		.num_packet = 1,
+		.packets = &enter_4_byte_cmd,
+		.timeout = INF_MSPI_S25H_DEFAULT_MSPI_TIMEOUT,
+	};
+
+	ret = mspi_transceive(config->bus, &config->dev_id, &xfer);
+	if (ret < 0) {
+		LOG_ERR("Error sending command to enter 4 byte address mode");
+		return ret;
+	}
+
+	data->mspi_dev_cfg.addr_length = 4;
+
```

Comment:
did you mean to do a `ret = flash_mspi_infineon_s25h_prepare_mspi_bus(dev);` here?